### PR TITLE
[zephyr] BootReasons and NetworkInterfaces fixes

### PR DESCRIPTION
**Pull request by Damian-Nordic**
1. Fix BootReasons attribute of GeneralDiagnostics cluster:
 - Make sure hwinfo_get_reset_cause() is called once per boot and is immediately followed by hwinfo_clear_reset_cause() since the reset cause is accumulated between subsequent resets.
 - Add support for SoftwareUpdateCompleted boot reason.
 - Reduce the scope of reasons classified as SoftwareReset.

2. Add a missing bound check to the NetworkInterfaces attribute provider and always assume that Zephyr's net_if API is available.

### CODE REVIEW

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp [hunk 1]
```diff
@@ -29,17 +29,76 @@
...
+
+    if (reason & RESET_BROWNOUT)
+    {
+        return BootReasonType::BrownOutReset;
+    }
+
+    if (reason & RESET_SOFTWARE)
+    {
+#ifdef CONFIG_MCUBOOT_IMG_MANAGER
+        if (mcuboot_swap_type() == BOOT_SWAP_TYPE_REVERT)
```
Comment (LuDuda) [inline src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:77]: I think this is a good enough, though i believe if another reset happens before we send NotifyApply then we won't send it anymore (at least immediately).

**Review by LuDuda — APPROVED** [review]
LGTM :+1:

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp (hunk 1 above)
Comment (Damian-Nordic) [reply src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:77]: But if another reset happens then MCUBoot will rollback the update, so it's good that we don't send NotifyUpdateApplied, right? We only need to make sure that OTA update will repeat within the next 24h.

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp [hunk 2]
```diff
@@ -29,17 +29,76 @@
...
+
 DiagnosticDataProviderImpl & DiagnosticDataProviderImpl::GetDefaultInstance()
 {
     static DiagnosticDataProviderImpl sInstance;
     return sInstance;
 }
 
+inline DiagnosticDataProviderImpl::DiagnosticDataProviderImpl() : mBootReason(DetermineBootReason())
+{
+    ChipLogDetail(DeviceLayer, "Boot reason: %" PRIu16, static_cast<uint16_t>(mBootReason));
```
Comment (kkasperczyk-no) [inline src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:99]: Why do we need to cast it and print as uint16_t?

**Review by kkasperczyk-no — APPROVED** [review]

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp [hunk 3]
```diff
@@ -29,17 +29,76 @@
...
+#endif
+
 #include <malloc.h>
 
 namespace chip {
 namespace DeviceLayer {
 
+namespace {
+
+uint8_t DetermineBootReason()
```
Comment (markaj-nordic) [inline src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:43]: Why not returning `BootReasonType`? And defining mBootReason as `BootReasonType` as well and only do a cast when printing?

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp [hunk 4]
```diff
@@ -215,16 +244,15 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetNetworkInterfaces(NetworkInterface **
         // Assuming IPv6-only support
         Inet::InterfaceAddressIterator interfaceAddressIterator;
         uint8_t ipv6AddressesCount = 0;
-        while (interfaceAddressIterator.HasCurrent())
+        while (interfaceAddressIterator.HasCurrent() && ipv6AddressesCount < kMaxIPv6AddrCount)
```
Comment (markaj-nordic) [inline src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:247]: Is this also related to boot reason?

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp (hunk 3 above)
Comment (markaj-nordic) [reply src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:43]: Ahh ok, `GetBootReason` accepts uint8_t.

**Review by markaj-nordic — APPROVED** [review]

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp (hunk 3 above)
Comment (kkasperczyk-no) [reply src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:43]: I don't remember precisely what was the problem, but I faced it some time ago. `BootReasonType` is defined in `DiagnosticDataProvider.h` and `GetBootReason` API is defined in the `ConfigurationManager.h`, where `DiagnosticDataProvider.h` cannot be included, so type is undefined. That's why uint8_t was used.
